### PR TITLE
SongSelect advanced filtering criteria

### LINKED ISSUE #3781: More in-game search queries
For a while now I've felt the need for more ways of sorting/listing beatmaps:

1.  Number of Objects ex. (objects>500) (o>500)
        -I feel like this would be efficient for picking out more dense maps, finding streamy maps to 
          practice. 
        -Also for finding maps that give the most score, for faster level ups.
2. Score achieved 
        -Especially for mania, this would be efficient for finding maps you can improve more on.
        -For someone who likes to go for million scores, they can easily see how many they have, how many they almost have.

3. Date Ranked
      -To separate older maps from newer maps.
      

Thanks

### PR DESCRIPTION
**Pull request by luaneko**
Will resolve #3781, but does not yet implement everything described, so don't merge yet.

The syntax is `<key><op><value>`, e.g. `stars:5`, `stars>3`, `stars<5`

Currently can filter by:

- Star difficulty
- Approach rate
- Drain rate
- Circle size
- Beat divisor
- Beatmap length
- ~~Object count~~
- Ranking status

### CODE REVIEW

**Comment by ticotaco72** [comment]
Wouldn't it be better to be done in UI?

File: osu.Game/Screens/Select/FilterCriteria.cs [hunk 1]
```diff
@@ -26,8 +27,46 @@ public string SearchText
...
             {
                 searchText = value;
-                SearchTerms = searchText.Split(',', ' ', '!').Where(s => !string.IsNullOrEmpty(s)).ToArray();
+                SearchTerms = searchText.Split(new[] { ',', ' ', '!' }, StringSplitOptions.RemoveEmptyEntries).ToArray();
             }
         }
+
+        public struct OptionalRange : IEquatable<OptionalRange>
+        {
+            public bool IsTrue(double value)
```
Comment (peppy) [inline osu.Game/Screens/Select/FilterCriteria.cs]: I'd probably call this function `IsInRange`.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
@phosphene47 I know it's been a while, but do you want to bring this up-to-date? Length has since been added so your database changes can be removed from this PR.

**Comment by phosphene47** [comment]
Sure. Would it be wiser to start a brand new PR from current master? I'm guessing there were a lot of changes since last December.

**Comment by peppy** [comment]
If you can get this PR in line, there's no need to make a new branch.

**Comment by peppy** [comment]
See #6161.

**Pull request closed without merging**